### PR TITLE
Add stats summary for scan results

### LINKED ISSUE #126: enhancement: --stats summary output (finding counts per rule and source)
You will add a `--stats` flag that prints a rollup: total findings, count per rule id, and for `scan --all`, count per source.

## How it fits together

```mermaid
flowchart TD
    A["cli.py<br/>_parse_run() (line 264)<br/>add --stats"] --> B["pipeline.run() (line 42)<br/>run_all() (line 291)"]
    B --> C["found_by_file / per_source"]
    C --> D["new _stats_payload()<br/>count per rule, per source"]
    D --> E["human: print a summary block"]
    D --> F["--json: nest under a stats key"]
```

Both output paths already hold the findings: `found_by_file` in `run()` and `per_source` in `run_all()`. Build the counts from those; do not rescan.

## Steps

1. Add `--stats` to `_parse_run()` in `cli.py` (line 264) and to `_get_completion_parser()` (line 502).
2. Write a helper in `pipeline.py` (for example `_stats_payload(found_by_file)`) that counts findings grouped by rule id. For `run_all()`, also group by source name.
3. Human output: print the summary after the findings table. With `--json`, nest the summary under a `"stats"` key so the existing findings schema keeps its shape for current consumers. Route JSON through `_emit_json_payload()` (line 1067).
4. Add a test: assert the counts match a fixture that plants a known mix across several rules and, for the `--all` case, several sources.

## Files

- `src/agentsweep/cli.py` — `_parse_run()` (line 264), `_get_completion_parser()` (line 502)
- `src/agentsweep/pipeline.py` — `run()` (line 42), `run_all()` (line 291), `_emit_json_payload()` (line 1067)
- `tests/` — new `test_stats.py`, or extend `tests/test_output.py`

## How to claim

Comment "I'll take this", fork the repo, branch from `main`, and open a PR that links this issue.

### PR DESCRIPTION
**Pull request by knight22-21**
## What & why

Adds a `--stats` flag for `scan`/`fix` that reports a findings rollup without rescanning existing results. In human-readable output it prints a summary after the findings table, and in `--json` mode it nests the rollup under a `stats` key so the existing `findings` list shape remains stable for existing consumers.

Fixes: #126 

## Type of change

- [ ] Bug fix
- [ ] New agent source
- [ ] New detection rule
- [x] Feature / enhancement
- [ ] Refactor / docs / chore

## Testing

Added focused coverage in `tests/test_stats.py` for:

- `--json --stats` output includes a top-level `stats` object while preserving the existing `findings` array
- Statistics aggregation for total findings and per-rule counts
- Per-source statistics when scanning with `--all`

```text
$ pytest tests/test_stats.py
=============================== test session starts ===============================
platform win32 -- Python 3.13.3, pytest-9.1.1, pluggy-1.6.0
collected 2 items

tests/test_stats.py ..                                                        [100%]

================================ 2 passed in 0.32s ================================
```

## Checklist

- [x] `pytest` passes locally, and CI is green on all platforms (Linux and Windows, Python 3.11–3.13)
- [ ] `mypy src/` is clean — it runs on every matrix leg and is the most common reason a PR goes red
- [x] No new runtime dependency, or the PR explains why one is needed (the core install is deliberately three packages)
- [x] No real secrets, tokens, or raw history-file contents are committed — tests use synthetic, non-live examples
- [x] Redaction / write-path changes preserve the corruption-prevention invariants (path containment, atomic replace, mandatory no-clobber backup, post-write validation, line-count preservation, audit trail)
- [ ] New detection rule? Added to `RULES` and `ROTATION_GUIDANCE` and a fixture/test, with every regex quantifier bounded (no unbounded `.*`), and the keyword pre-filter kept lossless
- [ ] New source? Registered in `SOURCES` with process markers, a menu entry, and a round-trip test (see `CONTRIBUTING.md` → "Adding a new source — checklist")
- [x] `--json` and non-TTY output stay machine-clean (no banners/styling, no UI import on those paths)
- [x] Did not re-introduce `force-include` in `pyproject.toml`

## Summary by CodeRabbit

- **New Features**
  - Added a `--stats` option to `scan` and `fix`, producing an overall findings summary with per-rule breakdown.
  - When using `--all`, stats also include per-source totals.
  - In JSON output modes (including reports), stats are now embedded into the emitted payloads; on TTY runs, stats are printed after findings.
  - If a scan finds no issues and an output file is provided, it now writes JSON containing both `findings: []` and the computed `stats`.

- **Tests**
  - Added deterministic coverage for clean scans with `--stats --output`, plus existing validation for JSON and human aggregated stats behavior.

<details><summary><h3>Greptile Summary</h3></summary>

This PR adds a `--stats` flag to `scan` and `fix` that computes a findings rollup (total count, per-rule breakdown, and per-source breakdown for `--all` runs) without requiring a rescan. JSON mode nests the rollup under a `stats` key alongside `findings`; human mode prints a stats block after the findings table; `--stats --format sarif` is explicitly rejected.

- `_stats_payload` / `_stats_payload_multi` aggregate findings counters from the already-scanned `found_by_file` structure; `_show_stats` renders them via `ui.console.print`.
- Clean-scan edge cases (no files, no sources, empty findings) are handled in both `run()` and `run_all()`, writing `{\"findings\": [], \"stats\": ...}` to `--output` when specified.
- Eight focused tests cover JSON wrapping, SARIF rejection, output-file writes, and multi-source rollup; the human-mode path test (`test_scan_human_stats_prints_summary`) uses `capsys` which cannot intercept the module-level Rich Console singleton.
</details>

<h3>Confidence Score: 4/5</h3>

The core stats computation and all JSON output paths are correct; the one gap is in the test suite where the human-mode coverage test will likely fail at assertion time.

The production code for _stats_payload, _stats_payload_multi, and _show_stats is logically sound, and JSON wrapping through _emit_json_payload is handled consistently. The only real defect is that test_scan_human_stats_prints_summary uses capsys against a module-level Rich Console singleton — that fixture cannot intercept the cached stdout reference, so all assertions on captured.out will fail, leaving the _show_stats / human-mode branch without effective test coverage.

**Files Needing Attention:** tests/test_stats.py — the human-mode stats test needs capfd instead of capsys.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/agentsweep/pipeline.py | Core stats implementation: adds _stats_payload, _stats_payload_multi, and _show_stats helpers; wires --stats through run() and run_all() for JSON/human modes; handles clean-scan edge cases and SARIF rejection. Two early-exit paths in run_all() call _emit_json_payload without return (discarding exit code, though functionally a no-op today). |
| src/agentsweep/cli.py | Adds --stats argument to both the main _parse_run parser and the completion sub-parsers for scan and fix; straightforward and consistent with existing flag patterns. |
| tests/test_stats.py | 8 new tests covering JSON+stats, SARIF rejection, output-file writes, and multi-source rollup. The human-readable path test (test_scan_human_stats_prints_summary) uses capsys, which cannot capture output from the module-level Rich Console singleton; those assertions will fail. Should use capfd instead. |

<sub>Reviews (8): Last reviewed commit: ["fix: emit stats object on empty-discover..."](https://github.com/ishannaik/agent-sweep/commit/8e1d13286c10417fc4ba3674a208ffd9a484020e) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=48248717)</sub>

### CODE REVIEW

**Review by Ishannaik — APPROVED** [review]
`--stats` wires into all three parsers. Clean-scan path prints stats before the early return. JSON output stays machine-readable. CI green 19/19 (mypy, Bandit, CodeQL, GitGuardian).

Two optional nits: trailing whitespace on the `cast` line in `_emit_json_payload`, and the `by_rule` loop duplicates between single/multi helpers. Neither blocks.

The JSON shape change (bare list vs wrapped object) is opt-in, so backward compat holds. Worth a release note.

We started a Discord for agent-sweep, CloakBin, and warp: https://discord.gg/KKvtRhQvRv
A star helps other people find the project.

**Comment by Ishannaik** [comment]
Good feature, two things to fix before it goes in.

**Empty discovery breaks the JSON contract.** I ran this on your branch:

```
$ agentsweep scan --root /tmp/empty --json --stats
No history files found under /tmp/empty
[]
```

Two problems in that output. The payload is a bare array with no `stats` key, so anything doing `payload["stats"]` crashes on an empty scan while working fine on every populated one. And the human line prints to stdout on a `--json` run, which breaks the machine-clean rule the repo holds everywhere else.

Same shape applies to the two early returns in `run_all` for `scan --all --json --stats`.

Fix is to emit the same object shape you already produce after a clean scan, with a zero stats block, and route that notice to stderr or drop it when `--json` is set.

**`--stats` is a silent no-op on the no-discovery path.** `scan --all --stats` with nothing on disk prints the stages and returns without calling `_show_stats`, so the flag does nothing rather than reporting zeros.

On tests: the human-output test asserts the rule names and `total findings: 3` but not the per-rule counts, so a wrong count line would still pass. Worth asserting the full lines and adding an empty-discovery `--json --stats` case.

The core work is solid, this is the edge case around it. https://discord.gg/KKvtRhQvRv if you want to discuss.

**Comment by Ishannaik** [comment]
Merged.

Verified on `8e1d132` before merging, since CodeRabbit had two Major flags open against the `--output` path and both are stale on this head:

- `scan --root <dir> --stats -o report.json` with findings writes `{"findings": [...], "stats": {...}}`, not findings alone.
- The zero-findings case writes the file too: `{"findings": [], "stats": {...}}`, exit 0. That was the "returns without creating the stats-bearing payload" flag.

621 passed, `mypy src/` clean on 28 files.

Greptile's P2 about `_show_stats` only being exercised through `--json --stats` still stands as a gap. Not blocking, but if you want a follow-up, a human-output assertion in `tests/test_stats.py` would close it.

File: src/agentsweep/pipeline.py [hunk 1]
```diff
@@ -166,6 +168,8 @@ def _print_empty_machine_output() -> None:
     total = sum(len(v) for v in found_by_file.values())
     ui.stage(3, "fail", "FINDINGS", f"{total} secret(s) in {len(found_by_file)} file(s)")
     _show_findings(found_by_file, source, output)
+    if getattr(args, "stats", False):
+        _show_stats(_stats_payload(found_by_file))
```
[the remark this answers is not in the document]
Comment (Ishannaik) [reply src/agentsweep/pipeline.py]: Not reproducing on the merged head. `scan --root <dir> --stats -o report.json` writes `{"findings": [...], "stats": {...}}`, so stats are on the `--output` path.

File: tests/test_stats.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+CODEX_LINE = (
+    '{"type":"message","role":"user","content":'
+    f'"token {GH_TOKEN}"' + '}\n'
+)
+
+
+@pytest.fixture(autouse=True)
+def _isolated_home(tmp_path, monkeypatch):
+    home = tmp_path / "home"
+    home.mkdir()
```
[the remark this answers is not in the document]
Comment (Ishannaik) [reply tests/test_stats.py:30]: Fair, and it stays open. Both tests go through `--json --stats`, so the human `_show_stats` rendering is still unasserted. Not merge-blocking; good follow-up issue.

File: src/agentsweep/pipeline.py [hunk 3]
```diff
@@ -157,6 +164,8 @@ def _print_empty_machine_output() -> None:
 
     if not found_by_file:
         ui.stage(3, "ok", "FINDINGS", "no secrets found")
+        if getattr(args, "stats", False):
+            _show_stats(_stats_payload(found_by_file))
```
[the remark this answers is not in the document]
Comment (Ishannaik) [reply src/agentsweep/pipeline.py]: Checked the zero-findings case specifically: exit 0 and the file is written as `{"findings": [], "stats": {...}}`. No early return.

File: tests/test_stats.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ishannaik) [reply tests/test_stats.py:30]: Correcting myself: this is already covered. `tests/test_stats.py:87` `test_scan_human_stats_prints_summary` calls `main(["scan", "--root", ..., "--stats"])` with no `--json` and asserts on the rendered text, including `total findings: 3`, the three `rule:` lines and `source:claude-code  3`. My earlier reply said the gap stood; it does not.
